sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala

```diff
@@ -250,7 +250,12 @@ object EnsureRequirements extends Rule[SparkPlan] {
 
   def apply(plan: SparkPlan): SparkPlan = plan.transformUp {
     // TODO: remove this after we create a physical operator for `RepartitionByExpression`.
-    case operator @ ShuffleExchangeExec(upper: HashPartitioning, child, _) =>
+    // SPARK-35989: AQE will change the partition number so we should retain the REPARTITION_BY_NUM
+    // shuffle which is specified by user. And also we can not remove REBALANCE_PARTITIONS_BY_COL,
+    // it is a special shuffle used to rebalance partitions.
+    // So, here we only remove REPARTITION_BY_COL in AQE.
+    case operator @ ShuffleExchangeExec(upper: HashPartitioning, child, shuffleOrigin)
+        if shuffleOrigin == REPARTITION_BY_COL || !conf.adaptiveExecutionEnabled =>
```

Comment:
Sorry for going back to here. After some more thought, it would be better to skip removing shuffle with all shuffle origin in AQE which include the `REPARTITION_BY_COL`. Then the condition can be changed to `if !conf.adaptiveExecutionEnabled`. Given this we can remove the final stage rules in `AdaptiveSparkPlanExec` safely which seems a little hack.

For `REPARTITION_BY_COL`, I want to explain it's also unsafe to remove. We may apply `OptimizeLocalShuffleReader` to change the output partitioning which does not follow the semantics of `RepartitionByExpression`.

What do you think ? @HyukjinKwon @cloud-fan

Reply:
This may introduce perf regression but can simplify AQE code a lot. What do you think? @maryannxue @JkSelf

Reply:
The `CoalesceShufflePartitions` rule may change the partition number when the `shuffleOrigin` is `REPARTITION_BY_COL`. I think it maybe a potential bug if removing the shuffle and not the perf issue.

Reply:
yeah, as we have only skipped applying `CoalesceShufflePartitions` or other custom shuffle reader at final stage. But for the stages which are in the process, we do nothing. That's why I think a little bit hack.

One other hack idea is we can remark the shuffle which is before the removed shuffle and change the `ENSURE_REQUIREMENTS` to `REPARTITION_BY_COL`. Then in AQE, we can do optimization safely.

IMO, I prefer the idea of `skip removing shuffle with all shuffle origin in AQE`, it's simple and it can be seen as a behavior change due to AQE is enabled by default. If user really hit this issue, they can just disable AQE.

Reply:
Changing sth. that would introduce regressions to the existing logic just to get the potential benefit of sth. we have not implemented yet is NOT acceptable.

A better solution here is to have a "root distribution (and potentially ordering) requirement", which will be passed to EnsureRequirement and ValidateRequirement, so that each custom-shuffle-reader rule (e.g., local-shuffle-reader, skew-join) can decide whether to apply just by calling `ValidateRequirement` or add back necessary shuffle by calling `EnsureRequirement`. There would be no need for final-stage hacks.

Reply:
given this, is it better to add a new physical plan for repartition ? which can aslo hold the `"root distribution (and potentially ordering) requirement"`. I remember some PR discussed about it before, and it should have other benefits, e.g. `AliasAwareOutputPartitioning`.

Reply:
This doesn't solve the real problem: shall we eliminate the new physical plan for user-specified shuffles and skip some AQE rules? Or do not eliminate and always run all the AQE rules?